### PR TITLE
Add Upload, Show and Download Artifacts Commands for all Apps

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR upstreams the artifact commands from the development from Lightning HPO.

```python
lightning upload artifacts --files tests/tests_lite/strategies/launchers
You are connected to the local Lightning App.
upload ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 100.0%
Your command execution was successful.
```

```python
lightning show artifacts
```

```python
You are connected to the local Lightning App.
📂 /drive/
📂 /drive/uploaded_files/
📂 /drive/uploaded_files/file_server/
📂 /drive/uploaded_files/file_server/tests/
📂 /drive/uploaded_files/file_server/tests/tests_lite/
📂 /drive/uploaded_files/file_server/tests/tests_lite/strategies/
📂 /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/
   /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/__init__.py
   /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/test_multiprocessing.py
   /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/test_subprocess_script.py
   /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/test_xla.py
📂 /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/__pycache__/
   /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/__pycache__/__init__.cpython-38.pyc
   /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/__pycache__/test_multiprocessing.cpy
thon-38-pytest-7.1.3.pyc
   /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/__pycache__/test_subprocess_script.c
python-38-pytest-7.1.3.pyc
   /drive/uploaded_files/file_server/tests/tests_lite/strategies/launchers/__pycache__/test_xla.cpython-38-pyte
st-7.1.3.pyc
Your command execution was successful.
```

```python
lightning download artifacts --output_dir=./output_dir
ls output_dir -> drive
```

TODOS:
* [] Add tests

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/lightning_app/api/http_methods.py [hunk 1]
```diff
@@ -29,7 +29,12 @@ def __init__(self, route: str, method: Callable, method_name: Optional[str] = No
             timeout: The time in seconds taken before raising a timeout exception.
         """
         self.route = route
-        self.component_name = method.__self__.name
+        self.is_function = not hasattr(method, "__self__")
```
Comment (lantiga) [inline src/lightning_app/api/http_methods.py:32]: this is unrelated to the PR right?

**Review by lantiga — COMMENTED** [review]
Hey @tchaton this makes sense in general, since apps have their shared_storage it is a good thing to give (some) users the ability to see and view data.
However we need to also have very strict control on who can actually run these commands. By default they should be limited to the app admin, but in order to have that we need to have app auth merged.
We could skateboard something where an env var (known to the admin) defines a token that is needed to access the API, although this would prevent Clone&Run from the UI.
I propose to wait for app auth to be merged and then activate this. In the meantime we can finalize and merge this, so one can opt-in use them, but let's avoid having the commands available in all apps by default.

File: src/lightning_app/api/http_methods.py (hunk 1 above)
Comment (tchaton) [reply src/lightning_app/api/http_methods.py:32]: This is required to enable function to be used instead of method of the flow.

**Comment by tchaton** [comment]
> Hey @tchaton this makes sense in general, since apps have their shared_storage it is a good thing to give (some) users the ability to see and view data. However we need to also have very strict control on who can actually run these commands. By default they should be limited to the app admin, but in order to have that we need to have app auth merged. We could skateboard something where an env var (known to the admin) defines a token that is needed to access the API, although this would prevent Clone&Run from the UI. I propose to wait for app auth to be merged and then activate this. In the meantime we can finalize and merge this, so one can opt-in use them, but let's avoid having the commands available in all apps by default.

Hey @lantiga, you are right ! Security is problematic there.

I have made the commands optional, but overall, we would need to come up with a general mechanism to enable the Rest API utilization, probably token protected.

Best,
T.C

**Comment by awaelchli** [comment]
My only concern is regarding the potentially blurred lines between what a dedicated CLI command is and what an app-specific command is. For example, one could have these as dedicated commands in the CLI:

lightning download data ...
lightning upload data ...

etc. 

Whereas here we are adding them to the app specific commands, i.e., you connect through `lightning connect` first. Initially, the app specific commands were intended to be custom commands provided by the user for the app, and not pre-defined ones from us. At least this is how I imagined it. 

As discussed offline briefly, an alternative could be to have the command classes as part of the framework, and the user can then choose to add them if they want.

**Comment by justusschock** [comment]
@tchaton is this still required?
